staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -595,8 +595,12 @@ func (c *Cacher) Watch(ctx context.Context, key string, opts storage.ListOptions
 	// underlying watchCache is calling processEvent under its lock.
 	c.watchCache.RLock()
 	defer c.watchCache.RUnlock()
-	watchRV = startWatchResourceVersionFn()
-	cacheInterval, err := c.watchCache.getAllEventsSinceLocked(watchRV)
+	forceAllEvents, err := c.waitUntilWatchCacheFreshAndForceAllEvents(ctx, requestedWatchRV, opts)
+	if err != nil {
+		return newErrWatcher(err), nil
+	}
+	startWatchRV := startWatchResourceVersionFn()
+	cacheInterval, err := c.watchCache.getAllEventsSinceLocked(startWatchRV, forceAllEvents)
```

Comment:
I suggest refactoring it to make it cleaner what exactly is happening as following:

In watchcache, extact the following function:
```
func (w *watchCache) getIntervalFromStoreLocked() (*watchCacheInterval, error) {
  ci, err := newCacheIntervalFromStore(w.resourceVersion, w.store, w.getAttrsFunc)
  if err != nil {
    return nil, err
  }
  return ci, nil
}
```

2. in `getAllEventsSinceLocked` function just call this function explicitly

3. Instead of passing `forceAllEvents` to `getAllEventsSinceLocked` change here to:
```
var cacheInterval *watchCacheInterval
if forceAllEvents {
  cacheInterval, err = c.watchCache.getIntervalFromStoreLocked()
} else {
  cacheInterval, err = c.watchCache.getAllEventsSinceLocked(startWatchRV)
}
```

For me that would be easier to follow with cleaner responsibilities.

Reply:
sure I can refactor.